pkg/nfspv/nfsvolume.go

```diff
@@ -0,0 +1,340 @@
+	if !isFinalizerRemoved {
+		// If finalizer is not deleted means finalizer doesn't exist so no need take action
+		return nil
+	}
+	_, err := n.clientset.CoreV1().
+		PersistentVolumes().
+		Update(context.TODO(), pvObj, metav1.UpdateOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to delete %s finalizer on PV %s", finalizer, pvObj.Name)
+	}
+	return nil
+}
+
+func removeFinalizer(objectMeta *metav1.ObjectMeta, finalizer string) bool {
+	originalFinalizers := objectMeta.GetFinalizers()
+
+	var finalizerCopy []string
+	var isFinalizerExist bool
+	for _, curFinalizer := range originalFinalizers {
+		if curFinalizer == finalizer {
+			isFinalizerExist = true
+			continue
+		}
+		finalizerCopy = append(finalizerCopy, curFinalizer)
+	}
+	objectMeta.Finalizers = finalizerCopy
+	return isFinalizerExist
+}
+
+func (n *nfsVolume) getVolumeData() (*NFSVolumeData, error) {
+	nfsPVC, err := n.getPVCCopy(n.pvObj.Spec.ClaimRef.Namespace, n.pvObj.Spec.ClaimRef.Name)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		// NotFound is a case where controller is down meantime user deleted NFS PVC
+		return nil, errors.Wrapf(err, "failed to get PVC %s/%s", n.pvObj.Spec.ClaimRef.Namespace, n.pvObj.Spec.ClaimRef.Name)
+	}
+
+	nfsPV, err := n.getPVCopy(n.pvObj.Name)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get PV %s", n.pvObj.Name)
+	}
+
+	// NOTE: We are naming backend PVC with "nfs-"+nfs-pv name
+	backendPVCName := "nfs-" + nfsPV.Name
+	backendPVC, err := n.getPVCCopy(n.nfsServerNamespace, backendPVCName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get PVC {%s/%s}", n.nfsServerNamespace, backendPVCName)
+	}
+
+	backendPV, err := n.getPVCopy(backendPVC.Spec.VolumeName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get PV %s", backendPVC.Spec.VolumeName)
+	}
+
+	return &NFSVolumeData{
+		NFSPVC:     nfsPVC,
+		NFSPV:      nfsPV,
+		BackingPVC: backendPVC,
+		BackingPV:  backendPV,
+	}, nil
+}
+
+func (n *nfsVolume) getPVCopy(pvName string) (*corev1.PersistentVolume, error) {
+	if n.pvObj.Name == pvName {
+		return n.pvObj.DeepCopy(), nil
+	}
+
+	pvObj, err := n.pvLister.Get(pvName)
+	if err != nil {
+		return nil, err
+	}
+	// Since we are fetching from cahce it is required
+	// to make deepcopy so that callers can mutuate object
+	return pvObj.DeepCopy(), nil
+}
+
+func (n *nfsVolume) getPVCCopy(namespace, name string) (*corev1.PersistentVolumeClaim, error) {
+	pvcObj, err := n.pvcLister.PersistentVolumeClaims(namespace).Get(name)
+	if err != nil {
+		return nil, err
+	}
+	// Since we are fetching from cahce it is required
+	// to make deepcopy so that callers can mutuate object
+	return pvcObj.DeepCopy(), nil
+}
+
+func (n *nfsVolume) isSupportedDataType() bool {
+	switch n.dataType {
+	case collectorinterface.JSONDataType:
+		return true
+	default:
+		return false
+	}
+}
+
+func getPatchData(oldObj, newObj interface{}) ([]byte, []byte, error) {
```

Comment:
how about creating a separate package for `getPatchData`, `removeFinalizer`, and other generic functions?

Reply:
Make sense!!

Reply:
Moved to helper package i.e `pkg/helper/helper.go`